### PR TITLE
Optimize KV cache dequantization performance

### PR DESCRIPTION
**Pull request by kyuyeunk**
This change reduces casting of quantized KV cache for best performance.

### CODE REVIEW

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py [hunk 1]
```diff
@@ -1954,12 +1954,14 @@ def masked_store(ref, val, start, end, group=1):
...
-      k = jnp.where(kv_mask, k.astype(jnp.float32), 0).astype(k.dtype)
-      v = jnp.where(kv_mask, v.astype(jnp.float32), 0).astype(v.dtype)
+      k = jnp.where(kv_mask, k, 0)
+      v = jnp.where(kv_mask, v, 0)
 
       qk = (
           jnp.einsum("nd,md->nm", q, k, preferred_element_type=jnp.float32) *
           sm_scale)
+      if k_scale is not None:
+        qk *= k_scale
```
Comment (yaochengji) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py]: It will become w8a8 instead of w8a16, right?

Comment (kyuyeunk) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py]: No, the query is still in 16 bits, so it's w8a16. But w8a8 feature is planned.

Comment (yaochengji) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py]: Thanks! I think we can multiply  k_scale and sm_scale first and then qk *= combined_scale. Thus the first one is a scalar multiplication.

Comment (kyuyeunk) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py]: I believe the compiler already takes care of this, but it's never a bad idea to be safe. Changed the code like you've suggested.

**Review by yaochengji — APPROVED** [review]
LGTM, thanks for the improvement!

**Comment by kyuyeunk** [comment]
> LGTM, thanks for the improvement!

Thanks! Can you press this PR's merge button for me?

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py [hunk 2]
```diff
@@ -1954,12 +1954,14 @@ def masked_store(ref, val, start, end, group=1):
       # kv lens will be contracting dim, we should mask out the NaNs.
       kv_mask = (
           lax.broadcasted_iota(jnp.int32, k.shape, 0) < kv_len - kv_len_start)
-      k = jnp.where(kv_mask, k.astype(jnp.float32), 0).astype(k.dtype)
-      v = jnp.where(kv_mask, v.astype(jnp.float32), 0).astype(v.dtype)
-
-      qk = (
-          jnp.einsum("nd,md->nm", q, k, preferred_element_type=jnp.float32) *
-          sm_scale)
+      k = jnp.where(kv_mask, k, 0)
```
Comment (yaochengji) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:1957]: Sorry, one more question: I'm not sure why k and v need to be converted to fp32 when applying the mask before the change. And I think this logic is not directly related to the kv cache quantization.

cc @bythew3i

**Comment by bythew3i** [comment]
> > LGTM, thanks for the improvement!
> 
> Thanks! Can you press this PR's merge button for me?

Please ping me if you check in any thing to RPA

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py (hunk 2 above)
Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:1957]: v5e does not support non-32 bit masking

Comment (kyuyeunk) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:1957]: Thanks for the clarification. If that's the case, I'll see if I can add branching depending on TPU type. I found conversiont of fp8 -> fp32 -> fp8 quite costly.

**Comment by kyuyeunk** [comment]
> > > LGTM, thanks for the improvement!
> > 
> > 
> > Thanks! Can you press this PR's merge button for me?
> 
> Please ping me if you check in any thing to RPA

ack. will always ping you on an RPA related changes.

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py (hunk 2 above)
Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:1957]: The v6e does not support your fp8 masking as well... anyhow it will convert to bf16 to do masking

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py [hunk 3]
```diff
@@ -2007,6 +2009,8 @@ def init_scratch_ref():
       m_curr = jnp.max(qk, axis=1, keepdims=True)
       s_curr = jnp.exp(qk - m_curr)
       qkv = jnp.dot(s_curr, v, preferred_element_type=jnp.float32)
+      if v_scale is not None:
```
Comment (bythew3i) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:2012]: Nice optimization!

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py (hunk 2 above)
Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:1957]: Please test the RPA in both v6e, v5e, and v4

Comment (kyuyeunk) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:1957]: got it. So is my understanding correct?
1. v6e supports bf16 masking while v5e only supports 32-bit masking (fp32).
2. v6e implicitly converts fp8 into bf16 so that's why above code worked. But in v5e, it does not & converting fp8 into fp32 has to be done explicitly.

Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:1957]: Yes, that is correct.

Actualy FP8 might work, I recall I added `byte_not_equal_to_zero`  op in Mosaic... but int4, bf16 might fail loudly when Mosaic does not handle the implicit conversion.

Comment (kyuyeunk) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:1957]: I've tested them out in v6e/v5e/v4 and verified that they are working correctly. It seems like implicit conversion works for older TPUs as well: http://sponge2/9329f606-b5ba-4e31-9d10-88554975d8df

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py (hunk 3 above)
Comment (kyuyeunk) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:2012]: thanks!

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py (hunk 2 above)
Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:1957]: Thanks for checking! I also just checked the Mosaic - actually it was me who added these implicit conversions to Mosaic in Jan .... man! my memory is getting so bad!

**Review by bythew3i — APPROVED** [review]
